File: 0155-min-stack/0155-min-stack.py
class MinStack:

    def __init__(self):
        self.st = []
        self.mn = None

    def push(self, val: int) -> None:
        if not self.st:
            self.st.append(val)
            self.mn = val
            
        else:
            if val <= self.mn:
                self.st.append(2*val-self.mn)
                self.mn = val
            else:
                self.st.append(val)

    def pop(self) -> None:
        if self.st:
            ele = self.st.pop()
            
            if ele <= self.mn:
                self.mn = (2*self.mn)-ele
            
    def top(self) -> int:
        if self.st:
            if self.st[-1] > self.mn:
                return self.st[-1] 
            else:
                ele = self.st[-1]
                return (ele+(2*self.mn-ele))//2

    def getMin(self) -> int:
        return self.mn


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 0155-min-stack/0155-min-stack.cpp
class MinStack {
private:
    stack<pair<int,int>> st;
public:    
    void push(int val) {
        if(!st.empty())
        {
            st.push(make_pair(val,min(st.top().second,val)));
        }
        else{
            st.push(make_pair(val,val));
        }
    }
    
    void pop() {
        if(!st.empty())
        {
            st.pop();
        }
    }
    
    int top() {
        if(!st.empty()){
            return st.top().first;
        }
        else{
            return INT_MIN;
        }
    }
    
    int getMin() {
        if(!st.empty()){
            return st.top().second;
        }
        else{
            return INT_MIN;
        }
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */